substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+
+	/// Next epoch authorities.
+	#[pallet::storage]
+	#[pallet::getter(fn next_authorities)]
+	pub type NextAuthorities<T: Config> = StorageValue<_, AuthoritiesVec<T>, ValueQuery>;
+
+	/// First block slot number.
+	///
+	/// As the slots may not be zero-based, we record the slot value for the fist block.
+	/// This allows to always compute relative indices for epochs and slots.
+	#[pallet::storage]
+	#[pallet::getter(fn genesis_slot)]
+	pub type GenesisSlot<T> = StorageValue<_, Slot, ValueQuery>;
+
+	/// Current block slot number.
+	#[pallet::storage]
+	#[pallet::getter(fn current_slot)]
+	pub type CurrentSlot<T> = StorageValue<_, Slot, ValueQuery>;
+
+	/// Current epoch randomness.
+	#[pallet::storage]
+	#[pallet::getter(fn randomness)]
+	pub type CurrentRandomness<T> = StorageValue<_, Randomness, ValueQuery>;
+
+	/// Next epoch randomness.
+	#[pallet::storage]
+	#[pallet::getter(fn next_randomness)]
+	pub type NextRandomness<T> = StorageValue<_, Randomness, ValueQuery>;
+
+	/// Randomness accumulator.
+	///
+	/// During block execution doesn't include randomness which ships within that block header.
+	#[pallet::storage]
+	pub type RandomnessAccumulator<T> = StorageValue<_, Randomness, ValueQuery>;
+
+	/// The configuration for the current epoch.
+	#[pallet::storage]
+	#[pallet::getter(fn config)]
+	pub type EpochConfig<T> = StorageValue<_, EpochConfiguration, ValueQuery>;
+
+	/// The configuration for the next epoch.
+	#[pallet::storage]
+	#[pallet::getter(fn next_config)]
+	pub type NextEpochConfig<T> = StorageValue<_, EpochConfiguration>;
+
+	/// Pending epoch configuration change that will be set as `NextEpochConfig` when the next
+	/// epoch is enacted.
+	///
+	/// In other words, a config change submitted during epoch N will be enacted on epoch N+2.
+	/// This is to maintain coherence for already submitted tickets for epoch N+1 that where
+	/// computed using configuration parameters stored for epoch N+1.
+	#[pallet::storage]
+	pub type PendingEpochConfigChange<T> = StorageValue<_, EpochConfiguration>;
+
+	/// Stored tickets metadata.
+	#[pallet::storage]
+	pub type TicketsMeta<T> = StorageValue<_, TicketsMetadata, ValueQuery>;
+
+	/// Tickets identifiers map.
+	///
+	/// The map holds tickets ids for the current and next epoch.
+	///
+	/// The key is a tuple composed by:
+	/// - `u8` equal to epoch's index modulo 2;
+	/// - `u32` equal to the ticket's index in a lexicographically sorted list of epoch's tickets.
+	///
+	/// Epoch X first N-th ticket has key (X mod 2, N)
+	///
+	/// Note that the ticket's index doesn't directly correspond to the slot index within the epoch.
+	/// The assigment is computed dynamically using an *outside-in* strategy.
+	///
+	/// Be aware that entries within this map are never removed, only overwritten.
+	/// Last element index should be fetched from the [`TicketsMeta`] value.
+	#[pallet::storage]
+	pub type TicketsIds<T> = StorageMap<_, Identity, (u8, u32), TicketId>;
```

Comment:
Can't this break when we skip epochs? This assumes that we always alternate between an odd and even epoch index, which might not be true if we skip epochs.

Reply:
See https://github.com/paritytech/polkadot-sdk/pull/1577#discussion_r1393093194